home/models.py

```diff
@@ -1268,3 +1269,251 @@ class PageView(models.Model):
+
+        # The name is needed for all templates to generate a name for the template
+        if not self.name:
+            errors.setdefault("name", []).append(
+                ValidationError("All WhatsApp templates need a name.")
+            )
+
+        message = self.message
+        # TODO: Explain what this does, or find a cleaner implementation
+        right_mismatch = re.findall(r"(?<!\{){[^{}]*}\}", message)
+        left_mismatch = re.findall(r"\{{[^{}]*}(?!\})", message)
+        mismatches = right_mismatch + left_mismatch
+
+        if mismatches:
+            errors.setdefault("message", []).append(
+                ValidationError(
+                    f"Please provide variables with matching braces. You provided {mismatches}."
+                )
+            )
+
+        vars_in_msg = re.findall(r"{{(.*?)}}", message)
+        non_digit_variables = [var for var in vars_in_msg if not var.isdecimal()]
+
+        if non_digit_variables:
+            errors.setdefault("message", []).append(
+                ValidationError(
+                    f"Please provide numeric variables only. You provided {non_digit_variables}."
+                )
+            )
+        # TODO: Add tests for these checks
+        # Check variable order
+        actual_digit_variables = [var for var in vars_in_msg if var.isdecimal()]
+        expected_variables = [str(j + 1) for j in range(len(actual_digit_variables))]
+        if actual_digit_variables != expected_variables:
+            errors.setdefault("message", []).append(
+                {
+                    "message": ValidationError(
+                        f'Variables must be sequential, starting with "{{1}}". You provided "{actual_digit_variables}"'
+                    )
+                }
+            )
+
+        if errors:
+            raise ValidationError(errors)
+
+        return result
+
+    @property
+    def fields(self):
+        """
+        Returns a tuple of fields that can be used to determine template equality
+        """
+        return (
+            self.category,
+            self.image,
+            self.message,
+            sorted(self.quick_reply_buttons),
+            self.example_values,
+        )
+
+    def create_whatsapp_template_name(self) -> str:
+        return f"{self.prefix}_{self.get_latest_revision().pk}"
+
+    def submit_whatsapp_template(self, previous_revision):
+        """
+        Submits a request to the WhatsApp API to create a template for this content
+
+        Only submits if the create templates is enabled
+        and if the template fields are different to the previous revision
+        """
+        if not settings.WHATSAPP_CREATE_TEMPLATES:
+            return
+
+        # If there are any missing fields in the previous revision, then carry on
+        if previous_revision:
+            previous_revision = previous_revision.as_object()
+            previous_revision_fields = previous_revision.whatsapp_template_fields
```

Comment:
I don't see any `whatsapp_template_fields` on the model, just `fields`, so maybe we want that instead?

Reply:
Yes we have renamed it back to `fields`

Reply:
Then I think this is only passing tests because we're never calling this function with a previous revision (it's always `submit_whatsapp_template(None)`), because I think this needs to be `previous_revision_fields = previous_revision.fields`

Reply:
When we make template submission its own action instead of something that happens when we save a revision, none of this comparison will need to happen anymore. I think it's fine to leave it broken for now and remove it entirely in the next PR.

Reply:
Sorry I missed this comment from Rudi after mine saying we renamed it back to `fields`